File: lib/data/recorder/audio_recorder_engine.dart
import 'dart:io';

import 'package:support_call_recorder/data/recorder/recorder_strategy.dart';

class AudioRecorderEngine {
  AudioRecorderEngine(this._strategies);
  final List<RecorderStrategy> _strategies;

  RecorderStrategy? _active;

  Future<String> start(File output) async {
    Object? lastError;
    for (final strategy in _strategies) {
      try {
        await strategy.start(output);
        _active = strategy;
        return strategy.sourceName;
      } catch (error) {
        lastError = error;
      }
    }
    throw StateError('No strategy could start: $lastError');
  }

  Future<void> stop() async {
    await _active?.stop();
    _active = null;
  }
}

File: app/src/main/java/com/reco/support/data/recorder/AudioRecorderEngine.kt
package com.reco.support.data.recorder

import android.content.Context
import java.io.File

interface RecorderStrategy {
    val name: String
    fun start(context: Context, targetFile: File)
    fun stop()
}

class AudioRecorderEngine(
    private val strategies: List<RecorderStrategy>
) {
    private var activeStrategy: RecorderStrategy? = null
    private var activeFile: File? = null

    fun start(context: Context, targetFile: File): String {
        var lastError: Throwable? = null
        for (strategy in strategies) {
            try {
                strategy.start(context, targetFile)
                activeStrategy = strategy
                activeFile = targetFile
                return strategy.name
            } catch (error: Throwable) {
                lastError = error
            }
        }
        throw IllegalStateException("No recorder strategy succeeded", lastError)
    }

    fun stop(): File? {
        activeStrategy?.stop()
        val output = activeFile
        activeStrategy = null
        activeFile = null
        return output
    }
}
